### PR TITLE
Add parser function to determine out of scope locations

### PR DESCRIPTION
**Pull request by ryanjduffy**
Associated Issue: #2964

### Summary of Changes

* Adds a new method to find out of scope locations
* Minor refactoring of parser helpers to use new `containsPosition` function

### Test Plan

* Unit tests!

### CODE REVIEW

**Comment by ryanjduffy** [comment]
Changed from my original approach of walking up the AST from a location to find scope boundaries to finding all functions and filtering from there. Both cases required an AST traversal (the former relied on an existing `getClosestScope`) but it was too difficult to "find" potential scopes in the original approach.

A possible solution to reducing AST traversals is to rewrite the functions that require AST searching to all run at the same time and cache relevant data. I've structured this function to adapt well to that in that it retrieves all functions (the cacheable data) and separates map/filter/reduces it to get to a result set based on a location. A similar approach could be taken with the other parser functions.

File: src/utils/parser/getOutOfScopeLocations.js [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+ */
+const getLocation = path => {
+  let location = Object.assign({}, get("node.loc", path));
+
+  // if the function has an identifier, start the block after it so the
+  // identifier is included in the "scope" of its parent
+  const identifierEnd = get("node.id.loc.end", path);
+  if (identifierEnd) {
+    location.start = identifierEnd;
+  }
```
Comment (jasonLaster) [inline src/utils/parser/getOutOfScopeLocations.js]: cute :)

**Comment by jasonLaster** [comment]
I switched to snapshots, which make it easier to detect changes in the output.

Here's a screenshot of how i verify the results. I think this is better, but I could be persuaded we want to cherry pick data out and make specific assertions. For now, this seems nice though

screen shot 2017-05-30 at 5 02 39 pm

**Comment by ryanjduffy** [comment]
👍

**Pull request merged**